askama_parser/src/expr.rs

```diff
@@ -709,60 +709,118 @@ impl<'a> Suffix<'a> {
-            let mut nested = 0;
-            let mut last = 0;
-            let mut in_str = false;
-            let mut escaped = false;
-
-            for (i, c) in input.char_indices() {
-                if !(c == '(' || c == ')') || !in_str {
-                    match c {
-                        '(' => nested += 1,
-                        ')' => {
-                            if nested == 0 {
-                                last = i;
-                                break;
-                            }
-                            nested -= 1;
-                        }
-                        '"' => {
-                            if in_str {
-                                if !escaped {
-                                    in_str = false;
-                                }
-                            } else {
-                                in_str = true;
-                            }
-                        }
-                        '\\' => {
-                            escaped = !escaped;
-                        }
-                        _ => (),
-                    }
-                }
+        #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+        enum Token {
+            SomeOther,
+            Open(Group),
+            Close(Group),
+        }
+
+        #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+        enum Group {
+            Paren,   // `(`
+            Brace,   // `{`
+            Bracket, // `[`
+        }
 
-                if escaped && c != '\\' {
-                    escaped = false;
+        impl Group {
+            fn as_close_char(self) -> char {
+                match self {
+                    Group::Paren => ')',
+                    Group::Brace => '}',
+                    Group::Bracket => ']',
                 }
             }
+        }
 
-            if nested == 0 {
-                let _ = input.next_slice(last);
-                Ok(())
-            } else {
-                fail.parse_next(input)
+        fn macro_arguments<'a>(i: &mut &'a str, open_token: Group) -> ParseResult<'a, Suffix<'a>> {
+            let start = *i;
+            let mut open_list: Vec<Group> = vec![open_token];
+            loop {
+                let before = *i;
+                let (token, token_span) = ws(opt(token).with_taken()).parse_next(i)?;
+                let Some(token) = token else {
+                    return Err(winnow::error::ErrMode::Cut(ErrorContext::new(
+                        "expected valid tokens in macro call",
+                        token_span,
+                    )));
+                };
+                let close_token = match token {
+                    Token::SomeOther => continue,
+                    Token::Open(group) => {
+                        open_list.push(group);
+                        continue;
+                    }
+                    Token::Close(close_token) => close_token,
+                };
+                let open_token = open_list.pop().unwrap();
```

Comment:
Are we sure there will always be at least one `open_list` item?

Reply:
Yes, the vector is initialized with one element, and `open_list.is_empty()` is the break condition.